### DISCUSSION TITLE
Is it possible to let Authelia pass Authorization header to protected app service?

### DISCUSSION THREAD

**Discussion by ilu123** [authelia/authelia#5445] [Q&A] [answered]
Our app service will automatically read **Authorization** header (which is Basic auth) of request when user login, after we integerate Authelia **2FA** auth into our Nginx, the Authorization header is gone, so even though the Authelia redirected the right URL to our app service, but the request has no Authorization header, so that it will show our app service login page again to user end.

So is it possible to let Authelia pass Authorization header to protected app service?

**Comment by james-d-elliott** [comment]
Authelia itself doesn't (and can't because of the architecture) remove any headers from the original  request by itself. You'll need to adjust your proxy configuration to handle this specific use case. Please show the related configurations and maybe someone can help you.

**Reply by ilu123** [reply]
I see. we're using LDAP as the Authelia's backend, and use the Authelia's login page for user's login by the "auth_request".
Well, I wonder if the Authelia can provide a Authorization header after user's Authelia login process, so that the user would not have to login again on the app's login page which will read that Authorization header if exists.

**Reply by james-d-elliott** [reply]
Technically it's possible but it's a incredibly serious bad security practice. The users password would have to be stored in clear text indefinitely. What we'd need to specifically help you are your proxy configurations for the app having the issue.

**Reply by ilu123** [reply]
yeah, well thanks, let me think about that.

**Reply by ilu123** [reply]
umm, after some research, finally I cannot find out a better way to achive the one-login page.
For now, after the Authelia's login for 2FA, the user has to login again on the App's login page. Authelia will lose Authorization header after 2FA success and then redirect to App's login page.....
In this opinion, Authelia seems not good for integerate into Existing App service, right?

**Reply by ilu123** [reply]
**my location.conf**

```
location /login {
satisfy all;
auth_ldap "Forbidden";
auth_ldap_servers myldap;  # which would generate the Authorization header for requests

    include /etc/nginx/authelia_proxy.conf;
    proxy_pass http://authelia:9091;
}

## Virtual endpoint created by nginx to forward auth requests.
location /authelia {
    ## Essential Proxy Configuration
    internal;

    proxy_pass http://authelia:9091/login/api/verify;

    ## Headers
    ## The headers starting with X-* are required.
    proxy_set_header X-Original-URL $scheme://$http_host$request_uri;
    proxy_set_header X-Original-Method $request_method;
    proxy_set_header X-Forwarded-Method $request_method;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_set_header X-Forwarded-Host $http_host;
    proxy_set_header X-Forwarded-Uri $request_uri;
    proxy_set_header X-Forwarded-For $remote_addr;
    proxy_set_header Content-Length "";
    proxy_set_header Connection "";
proxy_set_header            Proxy-Authorization $http_authorization;
    ## Basic Proxy Configuration
    proxy_pass_request_body off;
    proxy_next_upstream error timeout invalid_header http_500 http_502 http_503; # Timeout if the real server is dead
    proxy_redirect http:// $scheme://;
    proxy_http_version 1.1;
    proxy_cache_bypass $cookie_session;
    proxy_no_cache $cookie_session;
    proxy_buffers 4 32k;
    client_body_buffer_size 128k;

    ## Advanced Proxy Configuration
    send_timeout 5m;
    proxy_read_timeout 240;
    proxy_send_timeout 240;
    proxy_connect_timeout 240;
}

````

---

**auth.conf**

```
## Send a subrequest to Authelia to verify if the user is authenticated and has permission to access the resource.
auth_request /authelia;

proxy_set_header Host $http_host;


## Set the $target_url variable based on the original request.

## Comment this line if you're using nginx without the http_set_misc module.
set_escape_uri $target_url $scheme://$http_host$request_uri;

## Uncomment this line if you're using NGINX without the http_set_misc module.
# set $target_url $scheme://$http_host$request_uri;

## Save the upstream response headers from Authelia to variables.
auth_request_set $user $upstream_http_remote_user;
auth_request_set $groups $upstream_http_remote_groups;
auth_request_set $name $upstream_http_remote_name;
auth_request_set $email $upstream_http_remote_email;

## Inject the response headers from the variables into the request made to the backend.
proxy_set_header Remote-User $user;
proxy_set_header Remote-Groups $groups;
proxy_set_header Remote-Name $name;
proxy_set_header Remote-Email $email;

## Save the upstream authorization response headers from Authelia to variables.
                        auth_request_set $authorization $upstream_http_authorization;
                        auth_request_set $proxy_authorization $upstream_http_proxy_authorization;

                        ## Inject the authorization response headers from the variables into the request made to the backend.
                        proxy_set_header Authorization $authorization;
                        proxy_set_header Proxy-Authorization $proxy_authorization;


## If the subreqest returns 200 pass to the backend, if the subrequest returns 401 redirect to the portal.
error_page 401 =302 https://$http_host/login/?rd=$target_url;

```
---

**proxy.conf**

```
## Headers
proxy_set_header Host $http_host;
proxy_set_header X-Original-URL $scheme://$http_host$request_uri;
proxy_set_header X-Forwarded-Proto $scheme;
proxy_set_header X-Forwarded-Host $http_host;
proxy_set_header X-Forwarded-Uri $request_uri;
proxy_set_header X-Forwarded-Ssl on;
proxy_set_header X-Forwarded-For $remote_addr;
proxy_set_header X-Real-IP $remote_addr;
proxy_set_header Connection "";

## Basic Proxy Configuration
client_body_buffer_size 128k;
proxy_next_upstream error timeout invalid_header http_500 http_502 http_503; ## Timeout if the real server is dead.
proxy_redirect  http://  $scheme://;
proxy_http_version 1.1;
proxy_cache_bypass $cookie_session;
proxy_no_cache $cookie_session;
proxy_buffers 512 256k;

## Trusted Proxies Configuration
## Please read the following documentation before configuring this:
##     https://www.authelia.com/integration/proxies/nginx/#trusted-proxies
set_real_ip_from 223.255.255.0/24;
#set_real_ip_from 172.16.0.0/12;
#set_real_ip_from 192.168.0.0/16;
# set_real_ip_from fc00::/7;
real_ip_header X-Forwarded-For;
real_ip_recursive on;

## Advanced Proxy Configuration
send_timeout 5m;
proxy_read_timeout 360;
proxy_send_timeout 360;
proxy_connect_timeout 360;

```

**Reply by james-d-elliott** [reply, marked as answer]
Remove this section:

```nginx
## Save the upstream authorization response headers from Authelia to variables.
                        auth_request_set $authorization $upstream_http_authorization;
                        auth_request_set $proxy_authorization $upstream_http_proxy_authorization;

                        ## Inject the authorization response headers from the variables into the request made to the backend.
                        proxy_set_header Authorization $authorization;
                        proxy_set_header Proxy-Authorization $proxy_authorization;
```

**Comment by ilu123** [comment]
@james-d-elliott Maybe another question: I tested to setup `$http_authorization` to `Proxy-Authorization` by `proxy_set_header Proxy-Authorization $http_authorization;`
As far as my understanding, if I pass **Proxy-Authorization**, the Authelia should be able to read that header and skip the login page but to show the factor code input page, right? (for now, it seems not)

**Reply by james-d-elliott** [reply]
It will only perform 1FA, it's not valid for 2FA.

**Reply by james-d-elliott** [reply]
Considering the confused emoji, let me explain. The authorization headers are intended by web standards to communicate the complete authorization not an incomplete one. Users would have to include cookie headers in addition to the authorization header (and implement cookie management) for this to do 2FA, defeating the purpose of the Authorization header since they'd have to be redirected anyway. It's just not a pattern that has any reasonable benefits as far as I see it. 

We're obviously willing to hear some clearly articulated reasons why this would be desirable and to consider it however. Just wanting it "for undisclosed reasons" will not have any impact on our decision process however.

**Reply by ilu123** [reply]
@james-d-elliott  sorry for the emoji. Well let me explain a bit about my case.
My current services are: nginx, ldap, appXXX. We enabled nginx's auth_ldap feature, and our appXXX will read Authorization header to login. So when user access the appXXX, the browser will first popup basic auth dialog for user to input username & password. If pass, the appXXX will not show its login page again. If not, the user cannot access any resources. Just simple like that.

Now, we wanna enable/add 2FA feature to our services. And then we found Authelia. It really good for the ldap integeration feature and 2FA feature. But, we have to face a big problem that we have to let user login twice without modifying our appXXX login logic.
Am I right?

**Reply by james-d-elliott** [reply]
You are able to allow both basic scheme auth via the header for programmatic usage and use the user interaction browser portal flow. If the authorization header is absent the user is just forwarded to the auth portal (as long as they don't already have a valid session).

If the policy for a specific endpoint and user / network only requires the 1fa then the header will be sufficient. If it requires 2fa then it returns forbidden. So you could configure the default policy for app.example.com to be 2fa but allow users x y and z to do 1fa from either all ips or specific ips. This allows existing implementation to work almost identically to what it does now.

As far as the backend app authentication it is going to depend. Realistically are already known good users at that point. So you can skip auth or rely on the remote user header which includes their username.
